### PR TITLE
Improve validation of modules.yaml

### LINKED ISSUE #9857: Spack errors out if a syntactically wrong spec is in modules.yaml
Spack errors out if specs appearing in `modules.yaml` are syntactically invalid. On top of that the current reporting of the error is misleading on what should be done to fix the issue.

### Steps to reproduce the issue
Trying to use the following `modules.yaml`:
```yaml
modules:
  tcl:
    all:
      suffixes:
         '^python@2.7@': 'py2.7'
```
will cause **any** installation to fail. What should happen instead is that, upon installation, Spack:
1. Should give clear information on the issue ("syntactically wrong spec in file etc. ")
1. Should exit early (before starting the installation)

### Error Message
```console
$ spack install szip
==> Installing libszip
==> Searching for binary cache of libszip
==> Warning: No Spack mirrors are currently configured
==> No binary for libszip found: installing from source
==> Using cached archive: /home/mculpo/PycharmProjects/spack/var/spack/cache/libszip/libszip-2.1.1.tar.gz
==> Staging archive: /home/mculpo/PycharmProjects/spack/var/spack/stage/libszip-2.1.1-lmbbwwfqivit7uzhgucgt5ipr62d4fyq/szip-2.1.1.tar.gz
==> Created stage in /home/mculpo/PycharmProjects/spack/var/spack/stage/libszip-2.1.1-lmbbwwfqivit7uzhgucgt5ipr62d4fyq
==> No patches needed for libszip
==> Building libszip [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Error: SpecParseError: Invalid version specifier

/home/mculpo/PycharmProjects/spack/lib/spack/spack/package.py:1487, in build_process:
       1484                    echo = logger.echo
       1485                    self.log()
       1486
  >>   1487                # Run post install hooks before build stage is removed.
       1488                spack.hooks.post_install(self.spec)
       1489
       1490            # Stop timer.

See build log for details:
  /home/mculpo/PycharmProjects/spack/var/spack/stage/libszip-2.1.1-lmbbwwfqivit7uzhgucgt5ipr62d4fyq/szip-2.1.1/spack-build.out
```

### Information on your system

The only relevant detail is the `modules.yaml` above.

### PR DESCRIPTION
**Pull request by alalazo**
fixes #9857 

This PR improves the validation of `modules.yaml` by introducing a custom validator that checks if an attribute listed in `properties` or `patternProperties` is a valid spec. This new check applied to the test case in #9857 gives:
```console
$ spack install szip
==> Error: /home/mculpo/.spack/linux/modules.yaml:5: "^python@2.7@" is an invalid spec [Invalid version specifier]
```

@tgamblin @adamjstewart @scheibelp and others: can you please double check the schema? `modules.yaml` is fairly complex to check and more eyes could spot more bugs.

@trecakov Can you try this PR on the failure you reported in Slack? Hopefully you'll get an error report pointing to the exact line that needs to be changed.

### CODE REVIEW

File: lib/spack/spack/schema/modules.py [hunk 1]
```diff
@@ -9,110 +9,110 @@
...
-                        'items': {
-                            'type': 'string'
-                        }
+array_of_strings = {
+    'type': 'array', 'default': [], 'items': {'type': 'string'}
+}
+
+dictionary_of_strings = {
+    'type': 'object', 'patternProperties': {r'\w[\w-]*': {'type': 'string'}}
+}
```
Comment (alalazo) [inline lib/spack/spack/schema/modules.py]: @tgamblin I believe all comments have been addressed. Doing this refactor I saw a couple of minor things that you might want to do or not (either here or in another PR):

1. `array_of_strings` and `dictionary_of_strings` might be generic enough to be moved into `spack/schema/__init__.py` and be used in schema other than modules
2. `packages.yaml` could probably make use of `validate_spec` too

Let me know what do you think about these.

**Comment by alalazo** [comment]
This should be ready, at least for a further review.

**Comment by alalazo** [comment]
@tgamblin

File: lib/spack/spack/schema/modules.py [hunk 2]
```diff
@@ -127,8 +127,9 @@
                 'type': 'object',
                 'patternProperties': {
                     # prefix-relative path to be inspected for existence
-                    r'\w[\w-]*': {
-                        '$ref': '#/definitions/array_of_strings'}}},
+                    r'\w[\w-]*': array_of_strings
```
Comment (tgamblin) [inline lib/spack/spack/schema/modules.py]: Aaah. Isn't that better?

Comment (alalazo) [reply lib/spack/spack/schema/modules.py]: It's more readable indeed (but the schema can't be moved to a resource file external to Python code, which some time ago I thought it could have happened).

**Comment by alalazo** [comment]
I think I need some clarification on this comment:

> The only potential issue I see is the attempt to make regexes to validate Specs. The spec grammar isn't regular, so I think the best way to validate it is going to be to just parse a spec. I believe it is possible to do that during jsonschema validation, in much the same way the default setting validator currently works. That validator adds default values to the returned json data based on a default attribute. For this one, you could parse Specs and stick them in the data structure. I suspect something like that might also be useful in the spack.yaml (env) schema. We might need ways to ensure that they're written back the way they came in, though.

If we exclude attaching the parsed spec to the data structure, this PR was meant to introduce exactly these things. The part that does spec validation is:

https://github.com/spack/spack/blob/4ecb465a366de3f722aeb1525b22c0d22aefceae/lib/spack/spack/schema/__init__.py#L61-L73

and employs the spec parser to validate the spec and construct a meaningful error message (like the one in the description of the PR, pointing to the "Invalid version specifier"). The regex in the schema is just used to select the keywords in the configuration file on which we want to activate such validation.

TLDR: I feel I am missing the point of what you ask above :smile:

**Comment by hartzell** [comment]
I updated this AM and my `spack` command broke:

```console
[ghartzell@bifx1n03 spack]$ spack find
==> Error: 'tuple' object has no attribute 'get'
```

I narrowed it down to this commit and an interaction with my `modules.yaml`.

The following trimmed down `modules.yaml` file triggers the behavior.

```yaml
modules:
  enable::
    - lmod
  lmod:
    all:
      suffixes:
        '+jit': jit
        '^python@2.7': 'py2.7'
        '^python@3.6': 'py3.6'
```

I'm not sure what's changed and whether I need to update my `modules.yaml` files or if there's a rough bit in this change.

Help?

**Comment by alalazo** [comment]
@hartzell I can reproduce this even with the `modules.yaml` I used as a test to write the PR description:
```yaml
modules:
  tcl:
    all:
      suffixes:
         '^python@2.7@': 'py2.7'
```
so my best guess is that this PR is incompatible with something that got merged in the meanwhile and affects config files (but didn't trigger errors in the integration tests when rebasing :sob:).

**Comment by alalazo** [comment]
@tgamblin If this can wait a few days I'll work on a fix. Otherwise feel free to revert the PR, and I'll resubmit with a test and a fix - and apologies for the issue, was working at the time of submission and it passed tests at every rebase :cry:

**Comment by alalazo** [comment]
@hartzell @tgamblin I think I solved the issue. While rebasing a typo got in and there's an extra comma after the `module_file_configuration` dictionary. Fix in a sec.

**Comment by alalazo** [comment]
@hartzell Fixed

**Comment by hartzell** [comment]
That seems to be fixed, now I'm getting

```console
[ghartzell@bifx1n03 spack]$ spack find
==> Error: /home/ghartzell/spack/etc/spack/modules.yaml:28: Additional properties are not allowed ('environment_whitelist' was unexpected)
```

Here's the file in all of it's untrimmed glory:

```yaml
modules:
  enable::
    - lmod
  lmod:
    core_compilers:
      - 'gcc@4.8.5'
      - 'gcc@4.8.2'
      - 'gcc@4.8'
    # hash_length: 0
    whitelist:
      - gcc
    blacklist:
      - '%gcc@4.8.2'
      - '%gcc@4.8.5'
      - '%gcc@4.8'
    verbose_autoload: false
    all:
      suffixes:
        '+jit': jit
        '^python@2.7': 'py2.7'
        '^python@3.6': 'py3.6'
      filter:
        environment_blacklist: ['CPATH', 'LIBRARY_PATH', 'LD_LIBRARY_PATH']
    # WE DO NOT SET PACKAGE ROOT, it upsets GCC
    ^python:
      autoload:  'direct'
      filter:
        environment_whitelist: ['LD_LIBRARY_PATH']
```

**Comment by hartzell** [comment]
Sure enough, there's no mention of `environment_whitelist` in the source tree.  I wonder how long I've been carrying *that* around.  I think that it was supposed to override the blacklisting of `LD_LIBRARY_PATH` for things that depended on python....

**Comment by tgamblin** [comment]
> this PR was meant to introduce exactly these things. The part that does spec validation is: [...]

This is what i get for reviewing things on New Year's Eve.  I saw this:

https://github.com/spack/spack/blob/4ecb465a366de3f722aeb1525b22c0d22aefceae/lib/spack/spack/schema/modules.py#L12-21

and this:

https://github.com/spack/spack/blob/4ecb465a366de3f722aeb1525b22c0d22aefceae/lib/spack/spack/schema/modules.py#L109-114

Somehow I missed the `validate_spec: True` functionality.  and read the "needs to be updated" part as saying that some new expression needed to be updated to validate specs.  But it's really about the schema.  Sorry!

> TLDR: I feel I am missing the point of what you ask above

Apparently so am I!  😄 I think the only thing that might be worthwhile about my comment was the part about attaching the specs to the data structure, as it would avoid parsing specs twice.  But that's likely a minor optimization.  I guess adding spec validation to `env.py` would also be good, but that can happen later.

**Comment by hartzell** [comment]
> does what it says

Glad that I could provide an independent test case (aka "bad example").

Did I just make up the `environment_whitelist` or was that really a thing, once upon a time?

[Putting aside the "why" for a momement] Is there a way to *e.g.* prevent changes to `LD_LIBRARY_PATH` for everything *except* Python packages?

**Comment by hartzell** [comment]
> Is there a way to e.g. prevent changes to LD_LIBRARY_PATH for everything except Python packages?

Seems like @adamjstewart proposed a solution to this back in [this comment in my issue 3955](https://github.com/spack/spack/issues/3955#issuecomment-296385760).

Here's the example he provided:

```yaml
modules:
  lmod:
    all:
      filter:
        environment_blacklist: 'LD_LIBRARY_PATH'
    bowtie2:
      environment:
        LD_LIBRARY_PATH: '$prefix/lib'
```

I'll take it for a spin.

ps. *THIS* is why I love discussions in issue/PRs much much more than Slack, *etc*.  Thanks @adamjstewart!

**Comment by adamjstewart** [comment]
Tbh, I don't even remember that conversation lol. Let me add extra emphasis to **I haven't tested this, but it should work.** If it doesn't work, I would fully support the addition of an `environment_whitelist` key for symmetry reasons. I think the fact that you hallucinated the existence of this key is a pretty good argument for why it should exist.

**Comment by hartzell** [comment]
> the fact that you hallucinated the existence of this key is a pretty good argument for why it should exist.

That is a *very* dangerous precedent to be setting....

**Comment by alalazo** [comment]
> Is there a way to e.g. prevent changes to LD_LIBRARY_PATH for everything except Python packages?

I would have done something like this:
```yaml
modules:
  enable::
    - lmod
  lmod:
    core_compilers:
      - 'gcc@4.8.5'
      - 'gcc@4.8.2'
      - 'gcc@4.8'
    # hash_length: 0
    whitelist:
      - gcc
    blacklist:
      - '%gcc@4.8.2'
      - '%gcc@4.8.5'
      - '%gcc@4.8'
    verbose_autoload: false
    all:
      suffixes:
        '+jit': jit
        '^python@2.7': 'py2.7'
        '^python@3.6': 'py3.6'
      filter:
        environment_blacklist: ['CPATH', 'LIBRARY_PATH', 'LD_LIBRARY_PATH']
    # WE DO NOT SET PACKAGE ROOT, it upsets GCC
    ^python:
      autoload:  'direct'
      filter:
        environment_blacklist:: ['CPATH', 'LIBRARY_PATH']
```
where the important part is the double colon at the end (overriding of a property). That said same disclaimer as @adamjstewart : not tested but it should work.

**Comment by tgamblin** [comment]
@alalazo: is there going to be an inconsistency in how this is applied in `module refresh` vs. immediately after an install?  Immediately after install, I *believe* the build deps are still on the spec (unless you did something about that in the module hook). So this may allow things with build deps on python to change `LD_LIBRARY_PATH`.

**Comment by hartzell** [comment]
@alalazo's solution does not work on a refresh.

@adamjstewart's solution doesn't work because "someone" is doing a very careful validation of the `modules.yaml` file, which leads to:

```console
[ghartzell@bifx1n03 spack]$ spack module lmod refresh --delete-tree -y
==> Error: /home/ghartzell/spack/etc/spack/modules.yaml:34: Additional properties are not allowed ('LD_LIBRARY_PATH' was unexpected)
[ghartzell@bifx1n03 spack]$
```

:smiley: -> :scream: 

For completeness' sake, here's the `modules.yaml`:

```yaml
modules:
  enable::
    - lmod
  lmod:
    core_compilers:
      - 'gcc@4.8.5'
      - 'gcc@4.8.2'
      - 'gcc@4.8'
    # hash_length: 0
    whitelist:
      - gcc
    blacklist:
      - '%gcc@4.8.2'
      - '%gcc@4.8.5'
      - '%gcc@4.8'
    verbose_autoload: false
    all:
      suffixes:
        '+jit': jit
        '^python@2.7': 'py2.7'
        '^python@3.6': 'py3.6'
      filter:
        environment_blacklist: ['CPATH', 'LIBRARY_PATH', 'LD_LIBRARY_PATH']
    # WE DO NOT SET PACKAGE ROOT, it upsets GCC
    ^python:
      autoload:  'direct'
      # This doesn't exist (did it ever?)
      # filter:
      #   environment_whitelist: ['LD_LIBRARY_PATH']
      # This doesn't work, though alalazo thinks it should
      #  filter:
      #    environment_blacklist:: ['CPATH', 'LIBRARY_PATH']
      # This doesn't work, "Additional properties are not allowed..."
      # environment:
      #   LD_LIBRARY_PATH: '$prefix/lib'
```

**Comment by alalazo** [comment]
> solution doesn't work because "someone" is doing a very careful validation of the modules.yaml file, which leads to:

The check is indeed very careful :blush: You need an additional level that says what to do with the environment variable:
```yaml
environment:
  # This was missing above
  prepend:
    LD_LIBRARY_PATH: '$prefix/lib'
```
See [the tutorial](https://spack.readthedocs.io/en/latest/tutorial_modules.html#add-custom-environment-modifications) for an example.

**Comment by hartzell** [comment]
Wow, I'm an **excellent** bad example....

The suggestion of @alalazo + @adamjstewart, e.g.:

```yaml
      #...
      environment:
        prepend:
          LD_LIBRARY_PATH: '$prefix/lib'
```

leads to:

```console
[ghartzell@bifx1n03 spack]$ spack module lmod refresh --delete-tree -y
==> Error: /home/ghartzell/spack/etc/spack/modules.yaml:35: Additional properties are not allowed ('prepend' was unexpected)
[ghartzell@bifx1n03 spack]$
```

**Comment by alalazo** [comment]
@hartzell `prepend_path` :smile:

**Comment by adamjstewart** [comment]
> I wonder if we could easily tweak `jsonschema` to report which are the allowed properties.

PyTest does something like this and it's amazingly helpful:

```
E       fixture 'tmp_path' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, db_connection, doctest_namespace, monkeypatch, pytestconfig, record_xml_property, recwarn, tmpdir, tmpdir_factory
>       use 'pytest --fixtures [testpath]' for help on them.
```

**Comment by hartzell** [comment]
@alalazo -- `prepend_path` does not generate an error, but it *also* does not work.

I've been looking at the `py-pygments` Lua module file.  I also tried the `modules.yaml` file with `py-pygments` instead of `^python`, but neither worked.

It will (with `^python`) add an environment variable named `POODLE` (though `$prefix` doesn't do anything sensible, but that's solvable...)

Is this showing that the environment_blacklist overrides the specific settings?

**Comment by alalazo** [comment]
> Is this showing that the environment_blacklist overrides the specific settings?

Yes. The fact that we can't override on `environment_blacklist` is a bug to me though.
